File: demos/01-essentials/01-foundry/02-sdks/foundry-sdk-py/chat-model-router-multiturn.py
"""
Multi-turn Interactive Chat with Azure AI Model Router

This demo demonstrates:
- Interactive console-based chat session with conversation history
- Using Azure AI Foundry's model router to automatically select the best model for each turn
- Maintaining context across multiple turns with a rolling history window
- Displaying which underlying model was selected by the router for each response
- Configurable parameters (temperature, max tokens, history size) via environment variables

The model router intelligently selects from available models based on the request characteristics,
and the response.model property reveals which specific model handled each turn.
"""

import os
from typing import List, Dict
from azure.ai.projects import AIProjectClient
from azure.identity import DefaultAzureCredential
from dotenv import load_dotenv


def main():
    load_dotenv()

    # Load environment variables
    project_endpoint = os.environ["PROJECT_ENDPOINT"]
    router_model = os.environ["MODEL_ROUTER"]
    
    # System prompt can be customized via env
    system_prompt = os.getenv("SYSTEM_PROMPT", "You are a helpful assistant.")
    max_history_turns = int(os.getenv("MAX_HISTORY_TURNS", "12"))
    max_output_tokens = int(os.getenv("MAX_OUTPUT_TOKENS", "10000"))
    temperature = float(os.getenv("TEMPERATURE", "0.7"))
    top_p = float(os.getenv("TOP_P", "0.95"))
    frequency_penalty = float(os.getenv("FREQ_PENALTY", "0.0"))
    presence_penalty = float(os.getenv("PRES_PENALTY", "0.0"))

    project_client = AIProjectClient(
        endpoint=project_endpoint,
        credential=DefaultAzureCredential()
    )

    client = project_client.get_openai_client()

    # Conversation state: always begin with system message
    messages: List[Dict[str, str]] = [
        {"role": "system", "content": system_prompt},
    ]

    print("Interactive chat started. Type 'exit' or 'quit' to stop.\n")

    while True:
        try:
            user_input = input("You: ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\nExiting.")
            break

        if not user_input:
            continue
        if user_input.lower() in {"exit", "quit"}:
            print("Goodbye.")
            break

        # Append user turn
        messages.append({"role": "user", "content": user_input})

        # Trim history to keep context small while preserving the system message
        trimmed = trim_history(messages, max_history_turns=max_history_turns)

        # Call the router deployment
        response = client.chat.completions.create(
            messages=trimmed,
            max_tokens=max_output_tokens,
            temperature=temperature,
            top_p=top_p,
            frequency_penalty=frequency_penalty,
            presence_penalty=presence_penalty,
            model=router_model,
        )

        assistant_message = response.choices[0].message.content or ""
        routed_model = getattr(response, "model", "<unknown>")

        print("=" * 50)
        print(f"Model chosen by the router: {routed_model}")
        print("=" * 50)
        print(f"Assistant: {assistant_message}\n")

        # Append assistant turn to the full history
        messages.append({"role": "assistant", "content": assistant_message})


def trim_history(messages: List[Dict[str, str]], max_history_turns: int) -> List[Dict[str, str]]:
    """Keep system message plus the last N user/assistant turns.

    messages: full history including system at index 0
    max_history_turns: number of recent user+assistant pairs to retain
    """
    if not messages:
        return messages

    # Always preserve the first message if it is the system message
    head = messages[0:1]
    body = messages[1:]

    # Count pairs in body. We'll keep the last 2*max_history_turns messages
    keep = max_history_turns * 2
    if len(body) <= keep:
        return head + body
    else:
        return head + body[-keep:]


if __name__ == "__main__":
    main()

File: demos/01-essentials/01-foundry/02-sdks/extensions-ai/ChatModelRouterMultiturn.cs
using Microsoft.Extensions.AI;
using Azure.Identity;
using OpenAI;
using System.ClientModel.Primitives;
using ChatMessage = Microsoft.Extensions.AI.ChatMessage;

#pragma warning disable OPENAI001   // currently required for token-based authentication against Azure OpenAI

/// <summary>
/// Multi-turn Interactive Chat with Azure AI Model Router
/// 
/// This demo demonstrates:
/// - Interactive console-based chat session with conversation history
/// - Using Azure AI Foundry's model router to automatically select the best model for each turn
/// - Maintaining context across multiple turns with a rolling history window
/// - Displaying which underlying model was selected by the router for each response
/// - Configurable parameters (temperature, max tokens, history size) via appsettings.json
/// 
/// The model router intelligently selects from available models based on the request characteristics,
/// and the response Model property reveals which specific model handled each turn.
/// </summary>
public class ChatModelRouterMultiturn
{
    public async Task Run(
        string modelRouter, 
        string endpoint,
        string systemPrompt = "You are a helpful assistant.",
        int maxHistoryTurns = 12,
        int maxOutputTokens = 10000,
        float temperature = 0.7f,
        float topP = 0.95f,
        float frequencyPenalty = 0.0f,
        float presencePenalty = 0.0f)
    {
        // Create a BearerTokenPolicy using DefaultAzureCredential (Azure AD)
        var tokenPolicy = new BearerTokenPolicy(
            new DefaultAzureCredential(),
            "https://cognitiveservices.azure.com/.default");

        // Create the OpenAI ChatClient pointing at Azure OpenAI's OpenAI-compatible endpoint
        var chatClient = new OpenAI.Chat.ChatClient(
            model: modelRouter,
            authenticationPolicy: tokenPolicy,
            options: new OpenAIClientOptions
            {
                // Azure OpenAI compatible path
                Endpoint = new Uri(endpoint.TrimEnd('/') + "/openai/v1")
            }
        );

        // Adapt the OpenAI ChatClient to Microsoft.Extensions.AI's IChatClient
        IChatClient client = chatClient.AsIChatClient();

        // Conversation state: always begin with system message
        var messages = new List<ChatMessage>
        {
            new ChatMessage(ChatRole.System, systemPrompt)
        };

        Console.WriteLine("Interactive chat started. Type 'exit' or 'quit' to stop.\n");

        while (true)
        {
            Console.Write("You: ");
            var userInput = Console.ReadLine()?.Trim();

            if (string.IsNullOrEmpty(userInput))
            {
                continue;
            }

            if (userInput.Equals("exit", StringComparison.OrdinalIgnoreCase) ||
                userInput.Equals("quit", StringComparison.OrdinalIgnoreCase))
            {
                Console.WriteLine("Goodbye.");
                break;
            }

            // Append user turn
            messages.Add(new ChatMessage(ChatRole.User, userInput));

            // Trim history to keep context small while preserving the system message
            var trimmedMessages = TrimHistory(messages, maxHistoryTurns);

            // Prepare chat options
            var options = new ChatOptions
            {
                MaxOutputTokens = maxOutputTokens,
                Temperature = temperature,
                TopP = topP,
                FrequencyPenalty = frequencyPenalty,
                PresencePenalty = presencePenalty
            };

            // Call the router deployment
            ChatResponse response = await client.GetResponseAsync(trimmedMessages, options);

            var assistantMessage = response.Messages.FirstOrDefault(m => m.Role == ChatRole.Assistant)?.Text ?? "";
            var routedModel = response.ModelId ?? "<unknown>";

            Console.WriteLine("=" + new string('=', 49));
            Console.WriteLine($"Model chosen by the router: {routedModel}");
            Console.WriteLine("=" + new string('=', 49));
            Console.WriteLine($"Assistant: {assistantMessage}\n");

            // Append assistant turn to the full history
            messages.Add(new ChatMessage(ChatRole.Assistant, assistantMessage));
        }
    }

    /// <summary>
    /// Keep system message plus the last N user/assistant turns.
    /// </summary>
    /// <param name="messages">Full history including system at index 0</param>
    /// <param name="maxHistoryTurns">Number of recent user+assistant pairs to retain</param>
    /// <returns>Trimmed list of messages</returns>
    private List<ChatMessage> TrimHistory(List<ChatMessage> messages, int maxHistoryTurns)
    {
        if (messages.Count == 0)
        {
            return messages;
        }

        // Always preserve the first message if it is the system message
        var head = messages.Take(1).ToList();
        var body = messages.Skip(1).ToList();

        // Count pairs in body. We'll keep the last 2*maxHistoryTurns messages
        int keep = maxHistoryTurns * 2;
        
        if (body.Count <= keep)
        {
            return head.Concat(body).ToList();
        }
        else
        {
            return head.Concat(body.Skip(body.Count - keep)).ToList();
        }
    }
}
